Updated Continuous Aggregate Tutorial and some associated API fixes (#573)
timescale/docs.timescale.com-content: tutorials/continuous-aggs-tutorial.md

 min_trip_distance | numeric                     |           |          |
```

We can view the metadata for the continuous aggregate by selecting information 
from the **timescaledb_information.continuous_aggregates** view joined to the
**timescaledb_information.jobs** view.

``` sql
SELECT view_name, schedule_interval, 
  config ->> 'start_offset' as start_offset,
  config ->> 'end_offset' as end_offset,

>   date_trunc('second',next_start::timestamp) as next_start,

Author: yeah - I went back and forth. No reason too I guess - it was more of a nit-pick for me running through this. Can easily change back.

```

The column `job_id` gives the id of the background worker that updates the continuous
aggregate. `next_start` says when the next scheduled update will occur. 
`last_sucessful_finish` and `last_run_duration` allow you to see when the
scheduled aggregate finished and how long it took to run.

> ### 5. Update the Continuous Aggregate schedule

Reviewer: Add anchor tags for each of the steps?  They always end up coming in useful =)

Altering the schedule or parameters of a continuous aggregate policy is accomplished by
first removing the existing policy and then creating a new updated policy with the
desired settings.

To modify the policy that currently exists on the `cagg_rides_view` so that the `start_offset`
is now equal to `1 month`, you would:

``` sql

> SELECT remove_continuous_aggregate_policy('cagg_rides_view');

Reviewer: 1. Do you need to explicitly remove first before calling `add`?
1. You can't instead just `alter_job` ?  Or maybe there are 2 ways to do it?

Author: Well... this is where I admit that I didn't try `alter_job`. On a quick try it appears to work, however, I'm a bit unsure how much type/error checking we're doing. I mean, I was able to set "end_offset" = "chicken"... which although they have a pretty regular egg laying schedule, doesn't seem like a reasonable interval.

-[ RECORD 1 ]-------------------+---------------------------
view_name                       | cagg_rides_view
schedule_interval               | 00:30:00
start_offset                    | 1 week
end_offset                      | chicken
next_start                      | 2020-10-30 01:30:58
materialization_hypertable_name | _materialized_hypertable_9

It feels like the `remove`/`add` is the safer approach for now?

SELECT add_continuous_aggregate_policy('cagg_rides_view', 
  start_offset => interval '1 month',
  end_offset   => INTERVAL '1 hour',
  schedule_interval => INTERVAL '30 minutes');

SELECT view_name, schedule_interval, 
  config ->> 'start_offset' as start_offset,
  config ->> 'end_offset' as end_offset,
  date_trunc('second',next_start::timestamp) as next_start,
  materialization_hypertable_name
FROM timescaledb_information.continuous_aggregates ca
  INNER JOIN timescaledb_information.jobs j 
    ON ca.materialization_hypertable_name = j.hypertable_name
  WHERE view_name = 'cagg_rides_view';

-[ RECORD 1 ]-------------------+---------------------------
view_name                       | cagg_rides_view
schedule_interval               | 00:30:00

> start_offset                    | 1 mon

Reviewer: It's "1 mon" not "1 month"?

Author: yeah - surprised me too. Does that seem reasonable to others?

end_offset                      | 01:00:00
next_start                      | 2020-10-29 15:59:51
materialization_hypertable_name | _materialized_hypertable_9
```

> Notice that the `start_offset` is not set to `1 mon`.

Reviewer: ```suggestion
Notice that the `start_offset` is now set to `1 mon`.
```

## Using `WITH NO DATA` when creating a Continuous Aggregate [](with-no-data)

If you have a lot of historical data, we suggest creating the continuous aggregate
using the `WITH NO DATA` parameter for the `CREATE MATERIALIZED VIEW` command. Doing 
so will allow the continuous aggregate to be created instantly (you won't have to wait 
for the data to be aggregated on creation!). Data will then begin to populate as the 
continuous aggregate policy begins to run.

**However**, only data newer than `start_offset` would begin to populate the continuous
aggregate. If you have historical data that is older than the `start_offset` INTERVAL, 
you need to manually refresh history up to the current `start_offset` to allow
real-time queries to run efficiently.

Using the example in **Step 2** above, if we had years worth of data, a better approach 

> to creating the continuous aggregate is shown in the SQL below which includes `WITH NO DATA`
> and specifically refreshes the history separate from the continuous aggregate refresh policy:

Reviewer: ```suggestion
to creating the continuous aggregate is shown in the SQL below, which
1. Creates the continuous aggregate using the `WITH NO DATA`,
1. Refreshes the history explicitly, then
1. Adds the continuous aggregate refresh policy separately.
```

^ The above assumes you swap the order below.  Otherwise, you can swap back (ii) and (iii).

```sql
CREATE MATERIALIZED VIEW cagg_rides_view WITH
  (timescaledb.continuous)
AS
SELECT vendor_id, time_bucket('1h', pickup_datetime) as day,
  count(*) total_rides,
  avg(fare_amount) avg_fare,
  max(trip_distance) as max_trip_distance,
  min(trip_distance) as min_trip_distance
FROM rides
GROUP BY vendor_id, time_bucket('1h', pickup_datetime)
WITH NO DATA;

SELECT add_continuous_aggregate_policy('cagg_rides_view', 
  start_offset => INTERVAL '1 week',
  end_offset   => INTERVAL '1 hour',
  schedule_interval => INTERVAL '30 minutes');

> CALL refresh_continuous_aggregate('cagg_rides_view', NULL, localtimestamp - INTERVAL '1 week');

Reviewer: Wouldn't it be better to swap the order?  do the `CALL` first, then add the policy?

```

> You will find more details about the Continuous Aggregate API in the [documentation][cagg_api].

Reviewer: ```suggestion
You will find more details about the Continuous Aggregate API in its [documentation][cagg_api].
```

[hello_timescale]: /tutorials/tutorial-hello-timescale
[nyc_data]: https://timescaledata.blob.core.windows.net/datasets/nyc_data.tar.gz

> [tsdb_doc]: https://docs.timescale.com/v2.0/api#continuous-aggregates

Reviewer: 1. You don't need https://docs.timescale.com for a relative link
2. You don't need to specify version if you are always referring to the "same version" as switcher.  If user is on latest, it'll be latest.  If user is on /v2.0, it'll explicitly load /v2.0/api, etc.

```suggestion
[tsdb_doc]: /api#continuous-aggregates
```

Author: Yeah, sorry. This doesn't work with local hosting on relative links when you switch pages... so I keep forgetting this.
